Add wildcard option to metric-labels-allowlist

Mikulas opened the pull request:
**What this PR does / why we need it**:

https://github.com/kubernetes/kube-state-metrics/pull/1301 introduced backwards compatibility issue by dropping all labels not defined in `--metric-labels-allowlist`. That feature is great and should arguably be enabled for all new deployments. However, constructing such a list may prove quite hard for users migrating from previous versions, especially if kube-state-metrics is managed by a different team than the ones consuming eg pod label metrics.

This patch adds a new wildcard option that allows all labels on given resources. The assumption is that using this is a last resort when migrating to 2.0 (well, 2.1) from previous version. The suggested feature is as simple as possible to allow users to migrate. Generic pattern matching (regex) are explicitly a non-goal for this PR.

**Prior art:**
https://github.com/kubernetes/kube-state-metrics/pull/1289 allow metric labels by default
https://github.com/kubernetes/kube-state-metrics/pull/1247 PR with regex support

Fixes https://github.com/kubernetes/kube-state-metrics/issues/1309
Relates to https://github.com/kubernetes/kube-state-metrics/issues/1246

lilic commented:
@Mikulas Do you mind rebasing the PR, and we can have a look, thanks!

lilic commented on internal/store/pod_test.go:
```diff
@@ -1495,8 +1498,8 @@ func TestPodStore(t *testing.T) {
 	for i, c := range cases {
-		c.Func = generator.ComposeMetricGenFuncs(podMetricFamilies(nil))
-		c.Headers = generator.ExtractMetricFamilyHeaders(podMetricFamilies(nil))
+		c.Func = generator.ComposeMetricGenFuncs(podMetricFamilies(c.AllowLabelsList))
```
This is great but ideally, we want a separate test case for this. Keep testing what we did before and add this as a new test.

Mikulas replied:
Agreed, updated

lilic commented on docs/cli-arguments.md:
```diff
@@ -32,7 +32,7 @@ Usage of ./kube-state-metrics:
       --host string                       Host to expose metrics on. (default "::")
       --kubeconfig string                 Absolute path to the kubeconfig file
-      --labels-metric-allow-list string   Allows to pass a list of additional Kubernetes label keys that will be used in the resource' labels metric. By default the metric contains only name and namespace labels. To include additional labels provide a list of resource names in their plural form and Kubernetes label keys you would like to allow for them (Example: '=namespaces=[k8s-label-1,k8s-labe ...
+      --labels-metric-allow-list string   Allows to pass a list of additional Kubernetes label keys that will be used in the resource' labels metric. By default the metric contains only name and namespace labels. To include additional labels provide a list of resource names in their plural form and Kubernetes label keys you would like to allow for them (Example: '=namespaces=[k8s-label-1,k8s-labe ...
```
Just a thought: Wonder if there are scenarios where this might not work. 🤔 the argument having `*`. Like manifests in yaml or cli.

lilic commented on pkg/options/types.go:
```diff
@@ -130,6 +130,9 @@ func (n *NamespaceList) Type() string {
 	return "string"
 }
 
+// LabelWildcard allowlists any label
```
Do you mind adding a few test cases for the `TestLabelsAllowListSet` test in types_test.go file.

lilic reviewed (commented):
Nice work! Thank you! I am wondering if a single * would cause any problems, otherwise, this looks good apart from adding a few more test cases.

cc @tariq1890 @brancz please have a look as well as this changes our flags.

Mikulas commented:
Thanks for the review :) I rebased the PR and updated tests as recommended. I'd be happy to add more tests if deemed necessary but it seems like `pod_test.go` is the catchall for generic behaviour and the rest of stores only test resource specific behaviour.

Mikulas commented:
@lilic Please take a look when you get a chance :)

/assign lilic

lilic reviewed (approved):
/lgtm

I am sorry for the delay, I already reviewed but forgot to press the button 🤦‍♀️ 

Thank you so much for this!! 🎉

Mikulas commented:
Awesome, thank you :)
